sklearn/preprocessing/tests/test_polynomial.py

```diff
@@ -168,6 +194,109 @@ def test_spline_transformer_linear_regression(bias, intercept):
+    def f(x):
+        return np.sin(2 * np.pi * x) - np.sin(8 * np.pi * x) + 3
+
+    X = np.linspace(0, 1, 101)[:, None]
+    pipe = Pipeline(
+        steps=[
+            (
+                "spline",
+                SplineTransformer(
+                    n_knots=20,
+                    degree=3,
+                    include_bias=bias,
+                    extrapolation="periodic",
+                ),
+            ),
+            ("ols", LinearRegression(fit_intercept=intercept)),
+        ]
+    )
+    pipe.fit(X, f(X[:, 0]))
+
+    # Generate larger array to check periodic extrapolation
+    X_ = np.linspace(-1, 2, 301)[:, None]
+    predictions = pipe.predict(X_)
+    assert_allclose(predictions, f(X_[:, 0]), atol=0.01, rtol=0.01)
+    assert_allclose(predictions[0:100], predictions[100:200], rtol=1e-3)
+
+
+@pytest.mark.skipif(
+    sp_version < parse_version("1.0.0"),
+    reason="Periodic extrapolation not yet implemented for BSpline.",
+)
+def test_spline_transformer_periodic_spline_backport():
+    """Test that the backport of extrapolate="periodic" works correctly"""
+    X = np.linspace(-2, 3.5, 10)[:, None]
+    degree = 2
+
+    # Use periodic extrapolation backport in SplineTransformer
+    transformer = SplineTransformer(
+        degree=degree,
+        extrapolation="periodic",
+        knots=[[-1.0], [0.0], [1.0]]
+    )
+    Xt = transformer.fit_transform(X)
+
+    # Use periodic extrapolation in BSpline
+    coef = np.array([[1.0, 0.0], [0.0, 1.0], [1.0, 0.0], [0.0, 1.0]])
+    spl = BSpline(np.arange(-3, 4), coef, degree, "periodic")
+    Xspl = spl(X[:, 0])
+    assert_allclose(Xt, Xspl)
+
+
+def test_spline_transformer_periodic_splines_periodicity():
+    """
+    Test if shifted knots result in the same transformation up to permutation.
+    """
+    X = np.linspace(0, 10, 101)[:, None]
+
+    transformer_1 = SplineTransformer(
+        degree=3,
+        extrapolation="periodic",
+        knots=[[0.0], [1.0], [3.0], [4.0], [5.0], [8.0]]
+    )
+
+    transformer_2 = SplineTransformer(
+        degree=3,
+        extrapolation="periodic",
+        knots=[[1.0], [3.0], [4.0], [5.0], [8.0], [9.0]]
+    )
+
+    Xt_1 = transformer_1.fit_transform(X)
+    Xt_2 = transformer_2.fit_transform(X)
+
+    assert_allclose(Xt_1, Xt_2[:, [4, 0, 1, 2, 3]])
+
+
+@pytest.mark.parametrize("degree", [3, 5])
+def test_spline_transformer_periodic_splines_smoothness(degree):
+    """Test that spline transformation is smooth at first / last knot."""
+    X = np.linspace(0, 10, 1000)[:, None]
+
+    transformer = SplineTransformer(
+        degree=degree,
+        extrapolation="periodic",
+        knots=[[0.0], [1.0], [3.0], [4.0], [5.0], [8.0]]
+    )
+    Xt = transformer.fit_transform(X)
+
+    delta = (X.max() - X.min()) / len(X)
+    tol = 10
+
+    dXt = Xt
+
+    for d in range(1, degree + 1):
+        dXt = dXt[1:, :] - dXt[:-1, :]
+        assert (np.abs(dXt) < tol * (delta ** d)).all()
```

Comment:
```suggestion
    tol = 1e-3
    
    # continuity of function
    assert (np.abs(np.diff(Xt, axis=0)) < tol).all()
    
    # continuity of derivatives
    dXt = Xt
    for d in range(1, degree + 1):
        dXt = np.diff(dXt, axis=0) / delta  # d-th derivative
        # continuity of d-th derivative
        assert (np.abs(np.diff(dXt, axis=0)) < tol).all()
```
I hope this works. This is just a proposal to make this test cleaner.

Reply:
I'm now using `np.diff` and added a bunch of comments. Note that your proposal checks for the continuity of `Xt` twice, once on L292 and once in L299 for d = 1. Also `tol` needs to be of the order of (or rather bigger than) the expected size of derivatives. See the plot above. Since we're seeing values around 1, I've set `tol = 10`.

Reply:
I addressed some of this in https://github.com/scikit-learn/scikit-learn/pull/19483#issuecomment-788807414 below.